compiler/rustc_lint/src/if_let_rescope.rs

```diff
@@ -0,0 +1,306 @@
+    ///
+    /// ### Explanation
+    ///
+    /// With Edition 2024, temporaries generated while evaluating `if let`s
+    /// will be dropped before the `else` block.
+    /// This lint captures a possible change in runtime behaviour due to
+    /// a change in sequence of calls to significant `Drop::drop` destructors.
+    ///
+    /// A significant [`Drop::drop`](https://doc.rust-lang.org/std/ops/trait.Drop.html)
+    /// destructor here refers to an explicit, arbitrary implementation of the `Drop` trait on the type
+    /// with exceptions including `Vec`, `Box`, `Rc`, `BTreeMap` and `HashMap`
+    /// that are marked by the compiler otherwise so long that the generic types have
+    /// no significant destructor recursively.
+    /// In other words, a type has a significant drop destructor when it has a `Drop` implementation
+    /// or its destructor invokes a significant destructor on a type.
+    /// Since we cannot completely reason about the change by just inspecting the existence of
+    /// a significant destructor, this lint remains only a suggestion and is set to `allow` by default.
+    ///
+    /// Whenever possible, a rewrite into an equivalent `match` expression that
+    /// observe the same order of calls to such destructors is proposed by this lint.
+    /// Authors may take their own discretion whether the rewrite suggestion shall be
+    /// accepted, or rejected to continue the use of the `if let` expression.
+    pub IF_LET_RESCOPE,
+    Allow,
+    "`if let` assigns a shorter lifetime to temporary values being pattern-matched against in Edition 2024 and rewriting in `match` is an option to preserve the semantics up to Edition 2021",
+    @future_incompatible = FutureIncompatibleInfo {
+        reason: FutureIncompatibilityReason::EditionSemanticsChange(Edition::Edition2024),
+        reference: "issue #124085 <https://github.com/rust-lang/rust/issues/124085>",
+    };
+}
+
+declare_lint_pass!(
+    /// Lint for potential usages of async closures and async fn trait bounds.
+    IfLetRescope => [IF_LET_RESCOPE]
+);
+
+impl<'tcx> LateLintPass<'tcx> for IfLetRescope {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx hir::Expr<'tcx>) {
+        if !expr.span.edition().at_least_rust_2024() || !cx.tcx.features().if_let_rescope {
+            return;
+        }
+        let hir::ExprKind::If(cond, conseq, alt) = expr.kind else { return };
+        let hir::ExprKind::Let(&hir::LetExpr {
+            span,
+            pat,
+            init,
+            ty: ty_ascription,
+            recovered: Recovered::No,
+        }) = cond.kind
+        else {
+            return;
+        };
+        let expr_end = expr.span.shrink_to_hi();
+        let if_let_pat = expr.span.shrink_to_lo().between(init.span);
+        let before_conseq = conseq.span.shrink_to_lo();
+        let lifetime_end = {
+            let hi = conseq.span.hi() - BytePos(1);
+            conseq.span.with_hi(hi).shrink_to_hi()
+        };
```

Comment:
Problem: if we *have* to offset by one codepoint, please use [`SourceMap::end_point`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/source_map/struct.SourceMap.html#method.end_point) to compute span of the last character, the whitespace after `}`, instead of using `BytePos` manipulations. I suspect it's possible to hit the codepoint boundary assertion with a Unicode multi-byte whitespace via parser recovery or something.

Reply:
I'll double-check this logic a bit later, I find it weird that we have to chop off this whitespace manually.

Reply:
I believe that is a convention for a `Span` to be inclusive on the `start` and exclusive on the `end`. However, `SourceMap::end_point` has worked wonders for this purpose. Thank you so much for the suggestion! 🙇